tests/extra_source_and_tests/user_2d_T_shaped_pipe_VIPO_shell/T_shaped_pipe_VIPO_shell.cpp

```diff
@@ -0,0 +1,552 @@
+    Vecd(DL + BW, 2.0 * DH), Vecd(DL + BW, -DH), Vecd(DL1 - BW, -DH), Vecd(DL1 - BW, -BW), Vecd(-DL_sponge, -BW)};
+/** the inner wall polygon. */
+std::vector<Vecd> inner_wall_shape{
+    Vecd(-DL_sponge - BW, 0.0), Vecd(-DL_sponge - BW, DH), Vecd(DL1, DH), Vecd(DL1, 2.0 * DH + BW),
+    Vecd(DL, 2.0 * DH + BW), Vecd(DL, -DH - BW), Vecd(DL1, -DH - BW), Vecd(DL1, 0.0), Vecd(-DL_sponge - BW, 0.0)};
+//----------------------------------------------------------------------
+//	Define case dependent body shapes.
+//----------------------------------------------------------------------
+class WaterBlock : public MultiPolygonShape
+{
+  public:
+    explicit WaterBlock(const std::string &shape_name) : MultiPolygonShape(shape_name)
+    {
+        multi_polygon_.addAPolygon(water_block_shape, ShapeBooleanOps::add);
+    }
+};
+
+class ShellShape : public MultiPolygonShape
+{
+  public:
+    explicit ShellShape(const std::string &shape_name) : MultiPolygonShape(shape_name)
+    {
+        multi_polygon_.addAPolygon(outer_wall_shape, ShapeBooleanOps::add);
+        multi_polygon_.addAPolygon(inner_wall_shape, ShapeBooleanOps::sub);
+    }
+};
+
+/** Particle generator and constraint boundary for shell baffle. */
+class WallBoundary;
+template <>
+class ParticleGenerator<SurfaceParticles, WallBoundary> : public ParticleGenerator<SurfaceParticles>
+{
+    Real DL_sponge_;
+    Real resolution_ref_;
+    Real wall_thickness_;
+
+  public:
+    explicit ParticleGenerator(SPHBody &sph_body, SurfaceParticles &surface_particles,
+                               Real resolution_ref, Real wall_thickness)
+        : ParticleGenerator<SurfaceParticles>(sph_body, surface_particles),
+          DL_sponge_(20 * resolution_ref),
+          resolution_ref_(resolution_ref), wall_thickness_(wall_thickness){};
+    void prepareGeometricData() override
+    {
+        auto particle_number_mid_surface_01 = int((DL1 + DL_sponge_) / resolution_ref_);
+        //std::cout << " particle_number_mid_surface_01 = " << particle_number_mid_surface_01 << std::endl;
+        for (int i = 0; i < particle_number_mid_surface_01 - 1; i++)
+        {
+            Real x = -DL_sponge_ + (Real(i) + 0.5) * resolution_ref_;
+            // upper wall
+            Real y1 = DH + 0.5 * resolution_ref_;
+            addPositionAndVolumetricMeasure(Vecd(x, y1), resolution_ref_);
+            Vec2d normal_direction_1 = Vec2d(0, 1.0);
+            addSurfaceProperties(normal_direction_1, wall_thickness_);
+            // lower wall
+            Real y2 = - 0.5 * resolution_ref_; // lower wall
+            addPositionAndVolumetricMeasure(Vecd(x, y2), resolution_ref_);
+            Vec2d normal_direction_2 = Vec2d(0, -1.0);
+            addSurfaceProperties(normal_direction_2, wall_thickness_);
+        }
+
+        addPositionAndVolumetricMeasure(Vecd(DL1 - 0.5 * resolution_ref_, DH + 0.5 * resolution_ref_), resolution_ref_);
+        addSurfaceProperties(Vec2d(-1.0, 1.0), wall_thickness_);
```

Comment:
`Vec2d(-1.0, 1.0).normalized()` (same for the other corner)

Reply:
![rigid_shell](https://github.com/user-attachments/assets/a04885da-4461-4118-8c06-cd650c9d2302)
![rigid_shell02](https://github.com/user-attachments/assets/5f3600a9-22ff-4ba5-88c6-fb446d2150ad)
![rigid_shell03](https://github.com/user-attachments/assets/9fd880a9-e11d-4bbe-9901-dd2ed08a2540)
These images show the normal direction and average curvature of the rigid shell. After normalizing the normal vector at corners, flow instability still occurs.

Reply:
Thanks for the info, let me check if this problem exists in the velocity inlet flow